ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/RapidHeightMapExtractorCUDA.java

```diff
@@ -0,0 +1,499 @@
+   private float[] sensorToGroundTransformArray = new float[16];
+
+   private FloatPointer groundToSensorTransformHostPointer;
+   private FloatPointer groundToSensorTransformDevicePointer;
+
+   private FloatPointer sensorToGroundTransformHostPointer;
+   private FloatPointer sensorToGroundTransformDevicePointer;
+
+   private FloatPointer worldToGroundTransformHostPointer;
+   private FloatPointer worldToGroundTransformDevicePointer;
+
+   private FloatPointer groundToWorldTransformHostPointer;
+   private FloatPointer groundToWorldTransformDevicePointer;
+
+   private FloatPointer parametersHostPointer;
+   private FloatPointer parametersDevicePointer;
+
+   private CUDAKernel updateKernel;
+   private CUDAKernel registerKernel;
+   private CUDAKernel croppingKernel;
+   private CUDAKernel preprocessKernel;
+
+   private dim3 blockSize;
+   private dim3 preprocessKernelGridDim;
+   private dim3 updateKernelGridDim;
+   private dim3 registerKernelGridDim;
+   private dim3 croppingKernelGridDim;
+
+   private float[] paramsArray;
+
+   public RapidHeightMapExtractorCUDA(ReferenceFrame leftFootSoleFrame, ReferenceFrame rightFootSoleFrame)
+   {
+      footSoleFrames.put(RobotSide.LEFT, leftFootSoleFrame);
+      footSoleFrames.put(RobotSide.RIGHT, rightFootSoleFrame);
+   }
+
+   public void setDepthIntrinsics(CameraIntrinsics cameraIntrinsics)
+   {
+      this.cameraIntrinsics = cameraIntrinsics;
+   }
+
+   public void recomputeDerivedParameters()
+   {
+      centerIndex = HeightMapTools.computeCenterIndex(heightMapParameters.getLocalWidthInMeters(), heightMapParameters.getLocalCellSizeInMeters());
+      localCellsPerAxis = 2 * centerIndex + 1;
+      gridOffsetX = (float) heightMapParameters.getLocalWidthInMeters() / 2.0f;
+      globalCenterIndex = HeightMapTools.computeCenterIndex(heightMapParameters.getInternalGlobalWidthInMeters(),
+                                                            heightMapParameters.getInternalGlobalCellSizeInMeters());
+      globalCellsPerAxis = 2 * globalCenterIndex + 1;
+
+      cropCenterIndex = (heightMapParameters.getCropWindowSize() - 1) / 2;
+
+      if (2 * cropCenterIndex + 1 != heightMapParameters.getCropWindowSize())
+         throw new RuntimeException("The crop center index was computed incorrectly.");
+   }
+
+   public int getSequenceNumber()
+   {
+      return sequenceNumber;
+   }
+
+   public TerrainMapData getTerrainMapData()
+   {
+      return terrainMapData;
+   }
+
+   public Point3D getSensorOrigin()
+   {
+      return sensorOrigin;
+   }
+
+   public static HeightMapParameters getHeightMapParameters()
+   {
+      return heightMapParameters;
+   }
+
+   private void deallocateFloatPointer(FloatPointer hostPointer, Pointer devicePointer)
+   {
+      if (hostPointer != null)
+      {
+         hostPointer.close();
+         System.out.println("Deallocated host pointer.");
+      }
+
+      if (devicePointer != null)
+      {
+         devicePointer.close();
```

Comment:
If this is a CUDA pointer, you need to call `cudaFree` or `cudaFreeAsync` to actually deallocate the GPU memory